Adds a variant of fake handcuffs to the traitor uplink

ATH1909 opened the pull request:
## About The Pull Request

This PR adds a a special variant of the fake handcuffs toy to the traitor uplink as a 1 TC item. Unlike the toy version, these professional-grade fake handcuffs have the same name and description that normal handcuffs do, and the only noticeable difference between these fake handcuffs and "real" handcuffs is how long it takes to break out of the fake handcuffs: 1 second.

## Why It's Good For The Game

You can do some pretty neat things with fake handcuffs, but getting them is RNG dependent and requires some grinding at an arcade machine or spending a lot of money on toy crates.

You can, for example, handcuff yourself before Sec gets to you, then break out of your cuffs after the officer who brought you into Sec drops his guard (you can do the same thing with a freedom implant, but those things are expensive, yo). You can use these as normal handcuffs in a pinch if you really need some, as long as you keep your victim moving. You can do a Houdini act with them. You can "arrest" a fellow traitor, then reenact that scene from Episode IV with them. You can make secbots set to arrest mode and turrets set to stun mode stop attacking you/not attack you. Clown ops can quickly uncuff themselves if they accidentally cuff themselves with them. You can probably do more clever things with these than you can with a deck of syndicate playing cards, which costs 2 TC instead of 1 TC.

## Changelog
:cl: ATHATH
add: Adds professional-grade fake handcuffs to the traitor uplink.
balance: You can now handcuff yourself instantly and silently with fake handcuffs.
fix: One-handed clowns can no longer accidentally cuff themselves (successfully).
fix: A handcuffing noise is now made when a clown accidentally handcuffs themselves.
tweak: The feedback given when you fail to handcuff someone has been improved.
fix: Starting an attempt to break out of trick handcuffs no longer makes you unable to click on anything for 10 seconds, even after you finish breaking out of your trick handcuffs.
/:cl:

ATH1909 committed:
adds fake handcuffs to the traitor uplink

ATH1909 committed:
name change

ATH1909 committed:
Merge branch 'master' into fakehandcuffs

ATH1909 commented:
Some alternate versions of this PR that I had considered:
- Making this a clown/mime-only traitor item.
- Not adding this new item to the game or to the traitor uplink and instead just letting people print fake handcuffs from autolathes.

Do either of these ideas sound better to you?

Pumpkinoe commented:
I like the idea of it being an uplink item, but if you make it a clown/mime-only traitor item then it loses its surprise, because no sec officer with two brain cells is going to think "Oh wow, the clown's already cuffed, better bring him into sec!" without considering the clowncuffs.
I'd say keep it as an uplink item for all tator tots, because it sounds hilarious and fun.

Fikou commented:
ah yes, fake fake handcuffs

kriskog commented:
>You can probably do more clever things with these than you can with a deck of syndicate playing cards, which costs 2 TC instead of 1 TC.

Then why are these 1 TC 🤔

Pumpkinoe commented:
probably because I don't think anyone would buy them if they were 2 TC
maybe someone should reprice the syndicate playing cards since you can literally find those for free in certain parts of certain maps

PKPenguin321 commented:
the cards to this is like the revolver to the emag
you can do cooler stuff with an emag but it costs less because you cant just straight drop people with it

ATH1909 commented:
Well, fuck, this is apparently a thing: https://github.com/tgstation/tgstation/issues/48318

Should I just close this PR, then?

ATH1909 commented:
Should I make applying these cuffs to yourself (but not to others) instant?

ExcessiveUseOfCobblestone commented:
These types of items are what I like to see in the core uplink, much more than "It's a weapon but it DOES SOMETHING THAT NO OTHER WEAPON DOES: DAMAGES THEM WITH DIFFERENT FLAVOR!"

MrDoomBringer commented:
>Should I make applying these cuffs to yourself (but not to others) instant?
Yes, absolutely. Would make quickly cuffing youself in the middle of a chase possible.
 
Also, make resisting out of them give you the cuffs in-hand, rather than drop them on the floor. That way it's easier to re-use the fake handcuffs you spent good gamer money on.

ATH1909 committed:
fixes a bug and makes toy fake handcuffs a subtype of traitor fake handcuffs. also, you can cuff yourself instantly and silently with fake handcuffs now

ATH1909 commented:
Alright, I've added the instant (and silent) self-handcuffing part, but it looks like making them appear in your hands after you break out of them might be a bit too risky (bug-wise) for me to code (plus, it makes breaking out of them look more "genuine").

Toy fake handcuffs are now also a subtype of traitor fake handcuffs, so I (and future coders) don't/won't have to write the same procs for both of them.

Finally, I noticed that clumsy people could cuff themselves despite only having one arm, but only if they accidentally cuffed themselves while trying to cuff someone. I've added some more checks to apply_cuffs() that should hopefully fix that issue.

ATH1909 commented:
I still need to fix #48318...

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -146,9 +154,17 @@
+	if(C == user)
+		apply_cuffs(user, user) //instant self-application
+	else
+		..()
```
```suggestion
		apply_cuffs(user, user) //instant self-application
		return
	return ..()
```

ATH1909 committed:
Update code/game/objects/items/handcuffs.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

ATH1909 committed:
slightly modifies one comment

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -74,12 +74,20 @@
 				to_chat(user, "<span class='warning'>You fail to handcuff [C]!</span>")
 		else
-			to_chat(user, "<span class='warning'>[C] doesn't have two hands...</span>")
+			to_chat(user, "<span class='warning'>[C] doesn't have two hands...</span>") //if you're trying to cuff yourself while one-handed, this will refer to you in the third person, which is a bit weird, but not a big enough issue to write some shitcode to keep that from happening
```
```suggestion
			to_chat(user, "<span class='warning'>[C == user ? "You don't" : "[C] doesn't"] have two hands...</span>")
```

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -74,12 +74,20 @@
 	if(target.handcuffed)
 		return
 
+	if(!(target.get_num_arms(FALSE) >= 2 || target.get_arm_ignore()))
```
```suggestion
	if(!(target.get_num_arms(FALSE) >= 2) || target.get_arm_ignore())

```

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -74,12 +74,20 @@
 /obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = 0)
+
+	if(!istype(target)) //some sanity checks, in case apply_cuffs() gets called without attack() being called/fully run through (looking at YOU, clumsy people)
+		return
```
This shouldn't happen. Clumsy people are `carbon` as per the proc.
Avoid sanity checks that silently return. If something like this fails we want it to scream so we know where to fix.

ATH1909 replied:
A dextrous stand or a gorilla could click on themselves with fake handcuffs, no?

Rohesie replied:
It should produce no result:
```
/obj/item/restraints/handcuffs/attack(mob/living/carbon/C, mob/living/user)
	if(!istype(C))
		return
```

ATH1909 committed:
Update code/game/objects/items/handcuffs.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

ATH1909 committed:
lazy patch for the extra long click cooldown after breaking out of cuffs

ATH1909 commented:
Alright, I've made a lazy patch for the 10 second click cooldown applied whenever you try to break out of fake handcuffs. This isn't gonna affect anything else that invokes that 10 second click cooldown, just handcuffs (and unbuckling yourself from chairs and stuff while handcuffed).

If y'all are okay with it, I can change CLICK_CD_BREAKOUT to be 10 (deciseconds) instead of 100 (deciseconds). This would have the side effect of allowing people to start a new cuff breaking attempt once every second instead of once every 10 seconds, but I think that that's okay.

ATH1909 committed:
removes an unneeded sanity check

Rohesie commented on code/modules/uplink/uplink_items.dm:
```diff
@@ -1884,7 +1884,7 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	cost = 2
 	illegal_tech = FALSE
-
+  
```
```suggestion
 
```

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -74,10 +74,14 @@
 
 /obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = 0)
-	if(target.handcuffed)
+	if(target.handcuffed) //some sanity checks, in case apply_cuffs() gets called without attack() being called/fully run through (looking at YOU, clumsy people)
```
```suggestion
	if(target.handcuffed)
```
Comment no longer relevant, correct?

ATH1909 replied:
nah, target.handcuffed is still a sanity check, isn't it?

Rohesie replied:
What do clumsy people have to do with it?

ATH1909 replied:
They bypass the target.handcuffed check in attack(), so it needs to be re-checked here.

ATH1909 committed:
Update code/modules/uplink/uplink_items.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

ATH1909 committed:
improves the in-game feedback from cuffing attempts

ATH1909 commented:
Alright, I made some logic and feedback improvement changes so that it's more obvious to players why their cuffing attempts failed.

Also, accidentally cuffing yourself as a clown now makes a handcuffing sound (cuffing yourself with fake handcuffs still (intentionally) doesn't make a handcuffing sound, though).

ATH1909 committed:
Update handcuffs.dm

ATH1909 committed:
update for the descriptions of toy handcuffs

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -61,28 +61,39 @@
+		to_chat(user, "<span class='warning'>[C == user ? "You're" : "[C] is"] already handcuffed!</span>")
 
-/obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = 0)
+/obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = 0, intentional = 0)
```
```suggestion
/obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = FALSE, intentional = FALSE)

```

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -61,28 +61,39 @@
+			to_chat(user, "<span class='warning'>[target == user ? "You're" : "[target] is"] already handcuffed!</span>")
+		return
+
+	if(!(target.get_num_arms(FALSE) >= 2 || target.get_arm_ignore()))
```
```suggestion
	if(target.get_num_arms(FALSE) < 2 && !target.get_arm_ignore())
```

ATH1909 committed:
Update code/game/objects/items/handcuffs.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

ATH1909 committed:
Update code/game/objects/items/handcuffs.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

Rohesie commented on code/game/objects/items/handcuffs.dm:
```diff
@@ -61,28 +61,40 @@
+			to_chat(user, "<span class='warning'>[C == user ? "You don't" : "[C] doesn't"] have two hands...</span>")
+	else
+		to_chat(user, "<span class='warning'>[C == user ? "You're" : "[C] is"] already handcuffed!</span>")
+
+/obj/item/restraints/handcuffs/proc/apply_cuffs(mob/living/carbon/target, mob/user, dispense = FALSE, intentional = FALSE)
```
This will need documentation.

ATH1909 replied:
is the documentation that's here insufficient?

Rohesie replied:
The proc has no documentation.
Here's the guide:
https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

Rohesie commented on code/modules/uplink/uplink_items.dm:
```diff
@@ -1884,7 +1884,7 @@ GLOBAL_LIST_INIT(uplink_items, subtypesof(/datum/uplink_item))
 	cost = 2
 	illegal_tech = FALSE
-
+ 
```
```suggestion

```

ATH1909 committed:
Update code/modules/uplink/uplink_items.dm

Co-Authored-By: Rohesie <rohesie@gmail.com>

MrDoomBringer commented:
>:b:erge conflict

optimumtact commented:
Don't really see a need for this

ATH1909 commented:
The merge conflict resolver is being... "temperamental".

Probably because of the repo temporarily being down...

ExcessiveUseOfCobblestone commented:
26 days to fix documentation
